homeassistant/components/tplink/__init__.py

```diff
@@ -126,3 +186,29 @@ def legacy_device_id(device: SmartDevice) -> str:
     if "_" not in device_id:
         return device_id
     return device_id.split("_")[1]
+
+
+def _get_store(hass: HomeAssistant) -> Store[dict[str, dict[str, str]]]:
+    return Store[dict[str, dict[str, str]]](hass, STORAGE_VERSION, STORAGE_KEY)
+
+
+async def get_stored_credentials(hass: HomeAssistant) -> Optional[Credentials]:
+    """Retrieve the credentials from the Store."""
+    storage_data = await _get_store(hass).async_load()
+    if storage_data and (auth := storage_data[CONF_AUTHENTICATION]):
+        return Credentials(auth[CONF_USERNAME], auth[CONF_PASSWORD])
+
+    return None
```

Comment:
We should hold the storage object in `hass.data` when loading from `async_setup` for the life of the run so we don't have to load it every time

Reply:
Done